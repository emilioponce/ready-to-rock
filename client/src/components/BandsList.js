import React, { Component } from 'react'
import Band from './Band'
import Navigation from './Navigation'
import PropTypes from 'prop-types'
import { getBand } from '../api'

class BandsList extends Component {
  _isMounted = false

  constructor(props) {
    super(props)
    this.state = {
      current: 0,
      band: {}
    }
  }

  async componentDidMount() {
    this._isMounted = true
    const { current } = this.state
    const { bandsIds } = this.props
    // init with the first band of the id list
    const band = await getBand(bandsIds[current])
    if (this._isMounted) {
      this.setState({ band: band.data })
    }
  }

  async componentDidUpdate(prevProps, prevState) {
    const { bandsIds } = this.props
    const { current } = this.state
    // if band ids have change, then we need to show a new first band
    if (this.props.bandsIds !== prevProps.bandsIds) {
      const band = await getBand(bandsIds[current])
      this.setState({ current: 0, band: band.data })
    }
    // if current index has change, we request the new band
    if (this.state.current !== prevState.current) {
      const { current } = this.state
      const { bandsIds } = this.props
      const band = await getBand(bandsIds[current])
      this.setState({ band: band.data })
    }
  }

  componentWillUnmount() {
    this._isMounted = false
  }

  handleClick = e => {
    e.target.id === 'next'
      ? this.setState(state => ({ current: state.current + 1 }))
      : this.setState(state => ({ current: state.current - 1 }))
  }

  render() {
    const { current, band } = this.state
    const { bandsIds } = this.props
    return (
      <div>
        <Band title={band.title} body={band.body} members={band.members} />
        <Navigation
          current={current}
          size={bandsIds.length}
          handleClick={this.handleClick}
        />
      </div>
    )
  }
}

BandsList.propTypes = {
  bandsIds: PropTypes.array
}

export default BandsList
