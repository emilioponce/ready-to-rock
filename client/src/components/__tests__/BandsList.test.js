import React from 'react'
import ReactDOM from 'react-dom'
import BandsList from '../BandsList'
import { shallow } from 'enzyme'

const component = <BandsList bandsIds={[1, 2, 4, 5]} />

describe('Component BandsList', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div')
    ReactDOM.render(component, div)
    ReactDOM.unmountComponentAtNode(div)
  })

  it('snapshot renders correctly', () => {
    const wrapper = shallow(component)
    expect(wrapper).toMatchSnapshot()
  })

  it("Next value when 'next' button is clicked", () => {
    const wrapper = shallow(component)
    let Navigation = wrapper.find('Navigation')
    const handleClick = Navigation.props().handleClick
    const event = { target: { id: 'next' } }
    handleClick(event)
    Navigation = wrapper.find('Navigation')
    expect(Navigation.props().current).toBe(1)
  })

  it("Previous value when 'prev' button is clicked", () => {
    const wrapper = shallow(component)
    let Navigation = wrapper.find('Navigation')
    const handleClick = Navigation.props().handleClick
    const event = { target: { id: 'prev' } }
    handleClick(event)
    Navigation = wrapper.find('Navigation')
    // this value is no possible in the app, but this shows that decrements by 1
    expect(Navigation.props().current).toBe(-1)
  })
})
