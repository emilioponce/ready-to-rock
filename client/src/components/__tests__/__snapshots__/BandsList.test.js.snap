// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component BandsList snapshot renders correctly 1`] = `
<div>
  <Band />
  <Navigation
    current={0}
    handleClick={[Function]}
    size={4}
  />
</div>
`;
